include/storage/fbw.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for flash buffered writes
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_FBW_H_
+#define ZEPHYR_INCLUDE_STORAGE_FBW_H_
+
+/**
+ * @brief Abstraction over flash buffered writes.
+ *
+ * @defgroup fbw flash buffered write Interface
+ * @{
+ */
+
+#include <stdbool.h>
+#include <drivers/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*fbw_callback_t)(u8_t *buf, size_t len, size_t offset);
+
+/**
+ * @brief Structure for flash buffered writes
+ *
+ * Users should treat these structures as opaque values and only interact
+ * with them through the below API.
+ */
+struct fbw_ctx {
+	u8_t *buf; /* Write buffer */
+	size_t buf_len; /* Length of write buffer */
+	size_t buf_bytes; /* Number of bytes currently stored in write buf */
+	struct device *fdev; /* Flash device */
+	size_t bytes_written; /* Number of bytes written to flash */
+	size_t offset; /* Offset from base of flash device to write area */
+	size_t available; /* Available bytes in write area */
+	fbw_callback_t callback; /* Callback invoked on completed write op */
+#ifdef CONFIG_FBW_ERASE
+	off_t off_last; /* Last erased offset */
+#endif
+};
+
+/**
+ * @brief Initialize context needed for buffered write to flash.
+ *
+ * @param ctx context to be initialized
+ * @param fdev Flash device to operate on
+ * @param buf Write buffer
+ * @param buf_len Length of write buffer
+ * @param offset Offset within flash device
```

Comment:
Would it be correct to write this as:

```suggestion
 * @param offset Offset within flash device to start writing to
```

Reply:
Also, what kind of erase state does the sector containing this offset have to be in?

Reply:
The same as the other ones. If `FWB_ERASE=y` then they will be erased before written to. If `FWB_ERASE=n` then they have to be erased.